src/http/client.cc

```diff
@@ -127,7 +128,9 @@ input_stream<char> connection::in(reply& rep) {
 }
 
 future<> connection::close() {
-    return when_all(_read_buf.close(), _write_buf.close()).discard_result();
+    return when_all(_read_buf.close(), _write_buf.close()).discard_result().then([this] {
+        return std::move(_closed);
+    });
```

Comment:
Could it happen that move just copies _closed? lambda is not mutable so this ref can be const qualified, causing move cast to just copy _closed content rather than moving it to get rid of shared ptr ownership. I suggest, you make lambda mutable and use `std::exchange(_closed, make_ready_future<>()) ` instead to guarantee we're actually clearing this::_closed once it's consumed.

Reply:
We actually  delete future's copy constructor, unlike shared_future.

Could it happen we consume a const ref future, with _closed remaining unchanged?

Anyway I think it doesn't hurt to apply my suggestions.

Reply:
> Could it happen that move just copies _closed? 

No, future's copy constructor doesn't exist

> I suggest, you make lambda mutable and ...

It looks like non-mutability of a lambda doesn't make this->method() be const-only:

```c++
template <typename Fn>
static void static_fn(Fn&& f) {
    f();
}

struct foo {
    void fn() {
        static_fn([this] {
            fn2();
        });
    }

    void fn2() {
        fmt::print("non-const\n");
    }

    void fn2() const {
        fmt::print("const\n");
    }
};

int main() {
    foo f;
    f.fn();
}
```

prints "non-const"